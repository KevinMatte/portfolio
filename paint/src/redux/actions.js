

export const SESSION_LOGIN = "SESSION_LOGIN";
export const SESSION_LOGOUT = "SESSION_LOGOUT";
export const ADVANCED_MODE_TOGGLE = "ADVANCED_MODE_TOGGLE";

