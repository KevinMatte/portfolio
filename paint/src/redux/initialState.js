
export var initialState = {
    options: {
        advancedMode: false
    },
    session: {
        sessionId: null,
        permissions: [],
        title: "Kevin Matte's Portfolio",
    }
};
