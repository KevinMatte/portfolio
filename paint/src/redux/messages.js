/* Copyright (C) 2019 Kevin Matte - All Rights Reserved */

import {filterObject} from "../general/utils";
import {ReduxState} from "./reduxState";

export default class Messages extends ReduxState {
    constructor(initialState={
        list: [],
        messageId: 0,
        messageByField: {}
    }) {
        super(Messages, initialState);
    }

    static SESSION_MESSAGE = "SESSION_MESSAGE";

    static _add(dispatch, status, message, field = undefined) {

        if (typeof (message) !== 'string') {
            if (Array.isArray(message)) {
                message.forEach(submessage => Messages._add(dispatch, status, submessage));
            } else if (message.hasOwnProperty('message')) {
                let messageField = message.hasOwnProperty('field') ? message.field : undefined;
                let messageStatus = message.hasOwnProperty('status') ? message.status : undefined;
                Messages._add(dispatch, messageStatus, message.message, messageField);
            } else {
                return dispatch({
                    type: Messages.SESSION_MESSAGE,
                    message,
                    field,
                    status,
                });
            }
        } else {
            return dispatch({
                type: Messages.SESSION_MESSAGE,
                message,
                field,
                status,
            });
        }
    }

    static add = (status, message) => dispatch => {
        Messages._add(dispatch, status, message);
    };

    static getMessage = (message) => message ? message.message : null;

    static addReducer(messages, action) {
        let messageId = messages.messageId + 1;
        let {status, message, field} = action;
        let newMessage = {
            messageId,
            message,
            field,
            status,
        };
        let messageByField = messages.messageByField;
        if (field) {
            if (message)
                messageByField = {...messageByField, [field]: newMessage};
            else {
                messageByField = {...messageByField};
                delete messageByField[field];
            }
        }

        return {
            ...messages,
            messageId,
            list: Array.of(...messages.list, newMessage),
            messageByField,
        }
    }

    static MESSAGE_REMOVE_BY_FIELD = "MESSAGE_REMOVE_BY_FIELD";

    static removeByField(field) {
        return {
            type: Messages.MESSAGE_REMOVE_BY_FIELD,
            field,
        };
    }

    static removeByFieldReducer(messages, action) {
        let predicate;
        if (action.field.endsWith('/'))
            predicate = message => !message.field || !message.field.startsWith(action.field);
        else
            predicate = message => !message.field || message.field !== action.field;

        return {
            ...messages,
            list: messages.list.filter(predicate),
            messageByField: filterObject(messages.messageByField, predicate),
        };
    }

    static MESSAGE_REMOVE = "MESSAGE_REMOVE";

    static remove(messageId) {
        return {
            type: Messages.MESSAGE_REMOVE,
            messageId,
        };
    }

    static removeReducer(messages, action) {
        let predicate = message => message.messageId !== action.messageId;
        return {
            ...messages,
            list: messages.list.filter(predicate),
            messageByField: filterObject(messages.messageByField, predicate),
        };
    }

    static MESSAGE_TAKE = "MESSAGE_TAKE";

    static take(field) {
        return {
            type: Messages.MESSAGE_TAKE,
            field,
        };
    }

    static takeReducer(messages, action) {
        let predicate;
        if (action.field.endsWith('/'))
            predicate = message => !message.field || !message.field.startsWith(action.messageKey);
        else
            predicate = message => !message.field || message.field !== action.messageKey;

        return {
            ...messages,
            list: messages.list.filter(predicate),
        };
    }
}
