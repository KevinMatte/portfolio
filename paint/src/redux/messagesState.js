/* Copyright (C) 2019 Kevin Matte - All Rights Reserved */

import {filterObject} from "../general/utils";
import {ReduxState} from "./reduxState";

export default class MessagesState extends ReduxState {
    constructor(initialState={
        list: [],
        messageId: 0,
        messageByField: {}
    }) {
        super("MessagesState", initialState);
    }

    static _add(dispatch, status, message, field = undefined) {

        if (typeof (message) !== 'string') {
            if (Array.isArray(message)) {
                message.forEach(submessage => MessagesState._add(dispatch, status, submessage));
            } else if (message.hasOwnProperty('message')) {
                let messageField = message.hasOwnProperty('field') ? message.field : undefined;
                let messageStatus = message.hasOwnProperty('status') ? message.status : undefined;
                MessagesState._add(dispatch, messageStatus, message.message, messageField);
            } else {
                return dispatch({
                    type: MessagesState.MESSAGES_ADD,
                    message,
                    field,
                    status,
                });
            }
        } else {
            return dispatch({
                type: MessagesState.MESSAGES_ADD,
                message,
                field,
                status,
            });
        }
    }

    static MESSAGES_ADD = "MESSAGES_ADD";

    static add = (status, message) => dispatch => {
        this._add(dispatch, status, message);
    };

    static getMessage = (message) => message ? message.message : null;

    // noinspection JSUnusedGlobalSymbols
    static addReducer = (messages, action) => {
        let messageId = messages.messageId + 1;
        let {status, message, field} = action;
        let newMessage = {
            messageId,
            message,
            field,
            status,
        };
        let messageByField = messages.messageByField;
        if (field) {
            if (message)
                messageByField = {...messageByField, [field]: newMessage};
            else {
                messageByField = {...messageByField};
                delete messageByField[field];
            }
        }

        return {
            ...messages,
            messageId,
            list: Array.of(...messages.list, newMessage),
            messageByField,
        }
    }

    static MESSAGES_REMOVE_BY_FIELD = "MESSAGES_REMOVE_BY_FIELD";

    // noinspection JSUnusedGlobalSymbols
    static removeByField = (field) => {
        return {
            type: MessagesState.MESSAGES_REMOVE_BY_FIELD,
            field,
        };
    }

    // noinspection JSUnusedGlobalSymbols
    static removeByFieldReducer = (messages, action) => {
        let predicate;
        if (action.field.endsWith('/'))
            predicate = message => !message.field || !message.field.startsWith(action.field);
        else
            predicate = message => !message.field || message.field !== action.field;

        return {
            ...messages,
            list: messages.list.filter(predicate),
            messageByField: filterObject(messages.messageByField, predicate),
        };
    }

    static MESSAGES_REMOVE = "MESSAGES_REMOVE";

    // noinspection JSUnusedGlobalSymbols
    static remove = (messageId) => {
        return {
            type: MessagesState.MESSAGES_REMOVE,
            messageId,
        };
    }

    // noinspection JSUnusedGlobalSymbols
    static removeReducer = (messages, action) => {
        let predicate = message => message.messageId !== action.messageId;
        return {
            ...messages,
            list: messages.list.filter(predicate),
            messageByField: filterObject(messages.messageByField, predicate),
        };
    }

    static MESSAGES_TAKE = "MESSAGES_TAKE";

    // noinspection JSUnusedGlobalSymbols
    static take = (field) => {
        return {
            type: MessagesState.MESSAGES_TAKE,
            field,
        };
    }

    // noinspection JSUnusedGlobalSymbols
    static takeReducer = (messages, action) => {
        let predicate;
        let {messageKey} = action;
        if (action.field.endsWith('/'))
            predicate = message => !message.field || !message.field.startsWith(messageKey);
        else
            predicate = message => !message.field || message.field !== messageKey;

        return {
            ...messages,
            list: messages.list.filter(predicate),
        };
    }
}
