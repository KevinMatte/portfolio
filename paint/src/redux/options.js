/* Copyright (C) 2019 Kevin Matte - All Rights Reserved */

import {ReduxState} from "./reduxState";

export default class Options extends ReduxState {
    constructor(initialState={advancedMode: false}) {
        super(Options, initialState);
    }

    static ADVANCED_MODE_TOGGLE = "ADVANCED_MODE_TOGGLE";

    static toggleAdvancedMode() {
        return {
            type: Options.ADVANCED_MODE_TOGGLE,
        };
    }

    static toggleAdvancedModeReducer(options, /* action */) {
        return {
            ...options,
            advancedMode: !options.advancedMode,
        };
    }
}
