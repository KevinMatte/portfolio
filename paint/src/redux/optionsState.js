/* Copyright (C) 2019 Kevin Matte - All Rights Reserved */

import {ReduxState} from "./reduxState";

export default class OptionsState extends ReduxState {
    constructor(initialState={advancedMode: false}) {
        super(initialState);
    }

    static TOGGLE_ADVANCED_MODE = "TOGGLE_ADVANCED_MODE";

    // noinspection JSUnusedGlobalSymbols
    static toggleAdvancedMode() {
        return {
            type: this.TOGGLE_ADVANCED_MODE,
        };
    }

    // noinspection JSUnusedGlobalSymbols
    static toggleAdvancedModeReducer(options, /* action */) {
        return {
            ...options,
            advancedMode: !options.advancedMode,
        };
    }
}
