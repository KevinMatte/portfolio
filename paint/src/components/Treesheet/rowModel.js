import {getValueByPath} from "../../general/utils";
import {cloneObject} from "../../redux/pathUtils";

export default class RowModel {

    constructor(props) {
        this.props = props;
        this.sheetsByName = {};
        this.sheetNames = [];
        this.rows = [];
        this.numColumns = 0;
        this.columns = [];

        this._createRows(this.props.dataTree);
        this._updateSpreadsheetOpenRows();
    }

    _createRow(obj, path) {
        let {types} = this.props;

        let typeName = obj['type'];
        if (!typeName || !types.hasOwnProperty(typeName))
            typeName = obj['table'];
        let type = types[typeName];
        if (!type)
            return;

        let sheetName = `${typeName}_${path.length}`;

        let values = type.columns.map(column => {
            return getValueByPath(obj, column.path);
        });
        this.numColumns = Math.max(this.numColumns, type.columns.length);
        if (!this.sheetsByName.hasOwnProperty(sheetName)) {
            this.sheetsByName[sheetName] = {
                typeName,
                sheetName,
                path,
                type,
            };
            this.sheetNames.push(sheetName);
        }
        let row = {
            typeName,
            sheetName,
            values,
            path,
        };
        this.rows.push(row);

        if (type.hasOwnProperty('fields')) {
            type.fields.every((key) => {
                this._createRow(obj[key], [...path, key]);
                row.isOpen = true;
                return true;
            })
        }

        if (type.hasOwnProperty('arrays')) {
            type.arrays.every((key) => {
                this._createRows(obj[key], [...path, key]);
                row.isOpen = true;
                return true;
            })
        }
    }

    _createRows(aList, path = []) {
        aList.every((obj, iObj) => {
            this._createRow(obj, [...path, iObj]);
            return obj;
        });
    }

    duplicateRow(iFirstRow) {
        let path = this.rows[iFirstRow].path;
        this.props.duplicatePath(path);

        let iRow = iFirstRow;
        let iSplicePoint = iFirstRow;
        let newRows = [];
        let indent = this.rows[iRow].path.length;
        while (iRow < this.rows.length && (iRow === iFirstRow || this.rows[iRow].path.length > indent)) {
            let newRow = cloneObject(this.rows[iRow]);
            newRow.path[indent - 1]++;
            newRows.push(newRow);
            iRow++;
        }
        if (newRows.length > 0) {
            iSplicePoint = iRow;
            while (iRow < this.rows.length && this.rows[iRow].path.length >= indent) {
                this.rows[iRow].path[indent - 1]++;
                iRow++;
            }
            this.rows.splice(iSplicePoint, 0, ...newRows);
        }

        this._updateSpreadsheetOpenRows();
        return iSplicePoint;
    }

    deleteRow(iFirstRow) {
        let path = this.rows[iFirstRow].path;
        this.props.deletePath(path);

        let iRow = iFirstRow;
        let iSplicePoint = iFirstRow;
        let deleteCount = 0;
        let indent = this.rows[iRow].path.length;
        while (iRow < this.rows.length && (iRow === iFirstRow || this.rows[iRow].path.length > indent)) {
            iRow++;
            deleteCount++;
        }
        while (iRow < this.rows.length && this.rows[iRow].path.length >= indent) {
            this.rows[iRow].path[indent - 1]--;
            iRow++;
        }
        this.rows.splice(iSplicePoint, deleteCount);

        this._updateSpreadsheetOpenRows();
        return iSplicePoint;
    }

    toggleOpen(row) {
        row.isOpen = !row.isOpen;
        this._updateSpreadsheetOpenRows();
    }

    _updateSpreadsheetOpenRows() {
        let openIndent = -1;
        this.openRows = this.rows.filter((row) => {
            let indent = row.path.length;
            if (openIndent !== -1 && indent > openIndent)
                return false;
            if (indent === openIndent)
                openIndent = -1;
            if (!row.isOpen)
                openIndent = indent;
            return true;
        });
    }
}

