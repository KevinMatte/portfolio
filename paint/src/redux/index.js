/* Copyright (C) 2019 Kevin Matte - All Rights Reserved */

import thunk from 'redux-thunk';
import {createStore, applyMiddleware, compose} from 'redux';
import {combineReducers} from "redux";
import Options from './options';
import TempValues from './tempValues'
import Session from './session';
import Messages from './messages';
import Drawing from './drawing';

var rootReducer = combineReducers({
    session: Session.reducer,
    options: Options.reducer,
    tempValues: TempValues.reducer,
    messages: Messages.reducer,
    drawing: Drawing.reducer,
});


const middleWare = [
    thunk,
];

// noinspection JSUnresolvedVariable
const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;

const store = createStore(rootReducer, {}, composeEnhancers(
    applyMiddleware(...middleWare)
));

export default store;
