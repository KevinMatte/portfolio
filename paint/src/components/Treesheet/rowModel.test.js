import RowModel from "./rowModel";
import TreeState from "../../redux/treeState";
import {applyMiddleware, combineReducers, compose, createStore} from "redux";
import thunk from "redux-thunk";
import TempState from "../../redux/tempState";

let types = {
    drawing: {
        columns: [
            {label: 'Name', width: '150px', path: 'name'},
            {label: 'Type', width: '150px', path: 'type'},
        ],
        arrays: ['graphs'],
        fields: [],
    },
    graph: {
        columns: [
            {label: 'Name', width: '160px', path: 'name'},
            {label: 'Type', width: '160px', path: 'type'},
        ],
        arrays: ['points'],
        fields: [],
    },
    vector3: {
        columns: [
            {label: 'Name', width: '150px', path: 'name'},
            {label: 'Type', width: '150px', path: 'type'},
            {label: 'X', width: '150px', path: ['axis', 0]},
            {label: 'Y', width: '150px', path: ['axis', 1]},
            {label: 'Z', width: '150px', path: ['axis', 2]},
        ],
        array: [],
        fields: [],
    },
};
let models = [
    {
        name: 'First drawing',
        type: 'drawing',
        table: 'drawing',
        drawingId: 1,
        graphs: [
            {
                name: 'line 1',
                type: 'line',
                table: 'graph',
                points: [
                    {
                        type: 'point',
                        name: 'start',
                        table: 'vector3',
                        axis: [1, 2, 3],
                    },
                    {
                        type: 'point',
                        name: 'end',
                        table: 'vector3',
                        axis: [3, 2, 1],
                    },

                ]
            },
        ],
    },
];

const treeModel = new TreeState({types, models});
const rootReducer = combineReducers({
    treeModel: treeModel.reducer,
    tempValues: new TempState().reducer,
});

const middleWare = [
    thunk,
];

// noinspection JSUnresolvedVariable
const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;

const store = createStore(rootReducer, {}, composeEnhancers(
    applyMiddleware(...middleWare)
));

// noinspection ES6ModulesDependencies
test('RowModel', () => {

    let name = "treesheet";
    let state = store.getState();

    let props = treeModel.getMapDispatchToProps(store);
    props = {...props, name, types: state.treeModel.types, dataTree: state.treeModel.models};
    let model = new RowModel(props);
    expect(model.openRows).toMatchSnapshot();
    expect(model.dataTree).toMatchSnapshot();

    let oldRowsLength = model.openRows.length;
    let oldGraphsLength = state.treeModel.models[0].graphs.length;

    model.duplicateRow(1);
    expect(model.openRows.length).toBe(oldRowsLength + 3);
    state = store.getState();
    expect(state.treeModel.models[0].graphs.length).toBe(oldGraphsLength + 1);

    model.toggleOpen(model.rows[1]);
    expect(model.openRows.length).toBe(oldRowsLength + 1);

    // Do toggles.
    model.toggleOpen(model.rows[1]);
    expect(model.openRows.length).toBe(oldRowsLength + 3);
    model.toggleOpen(model.rows[0]);
    expect(model.openRows.length).toBe(1);
    model.toggleOpen(model.rows[0]);
    expect(model.openRows.length).toBe(oldRowsLength + 3);

    model.deleteRow(1);
    expect(model.openRows.length).toBe(oldRowsLength);
    state = store.getState();
    expect(state.treeModel.models[0].graphs.length).toBe(oldGraphsLength);
});