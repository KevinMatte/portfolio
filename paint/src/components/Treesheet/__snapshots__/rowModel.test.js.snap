// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RowModel 1`] = `
Array [
  Object {
    "isOpen": true,
    "path": Array [
      0,
    ],
    "sheetName": "drawing_1",
    "typeName": "drawing",
    "values": Array [
      "First drawing",
      "drawing",
    ],
  },
  Object {
    "isOpen": true,
    "path": Array [
      0,
      "graphs",
      0,
    ],
    "sheetName": "graph_3",
    "typeName": "graph",
    "values": Array [
      "line 1",
      "line",
    ],
  },
  Object {
    "path": Array [
      0,
      "graphs",
      0,
      "points",
      0,
    ],
    "sheetName": "vector3_5",
    "typeName": "vector3",
    "values": Array [
      "start",
      "point",
      1,
      2,
      3,
    ],
  },
  Object {
    "path": Array [
      0,
      "graphs",
      0,
      "points",
      1,
    ],
    "sheetName": "vector3_5",
    "typeName": "vector3",
    "values": Array [
      "end",
      "point",
      3,
      2,
      1,
    ],
  },
]
`;

exports[`RowModel 2`] = `undefined`;
